rfcs/text/0408-versioning-types.md

```diff
@@ -0,0 +1,355 @@
+- 'Constraints' refer to all JSON schema keywords which affect the validation of data in the Block Protocol, these include (but are not limited to) `type`, `properties`, `minItems`, `maxItems`, `minimum`, `maximum`
+- 'Semantic Annotations' refer to all JSON schema keywords which do not affect the validation of data in the Block Protocol, these include (but are not limited to) `title`, `description`, `examples`, `default`
+- The ordering of constraints does not affect the compatibility of schemas
+
+### Data Types
+
+At the present data types represent completely disjoint value spaces, and the base primitive types are all **incompatible** with one another. As the [Non-Primitive Data Types RFC](https://github.com/blockprotocol/blockprotocol/pull/355) continues to be specified, it should comment on how checking compatibility will be affected.
+
+### Property Types
+
+1.  Due to how property types are used, they implicitly affect the structure of the data in that they require the existence of a key that matches their base URI.
+1.  Property types define their constraints through a `oneOf` field. This `oneOf` is a collection of one or more values where each value is a variant of the "Property Values" (defined below).
+
+A property type `A` is therefore compatible with another property type `B` if and only if
+
+1.  they have the same base URI (which implies they are the same property type but perhaps different versions) and,
+1.  every element of `A`'s `oneOf` is compatible with one element of `B`'s `oneOf`
+
+The "Property Values" is a recursive definition which refers to the sub-schemas defined within the [property types meta-schema](https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0352-graph-type-system.md#property-types-1). It defines the element as being one of the following:
+
+#### Data Type Reference
+
+A data type reference is an inline reference to a data-type. A data type reference will be compatible with another data type reference if the data types that are referred to are also compatible.
+
+> _Without_ non-primitive data types this implies that the references are to the same data type.
+
+#### Property Type Object
+
+A poperty type object is JSON object defining a schema which is used as `oneOf` the permitted values for a property type, where the schema is of `type: "object`, and within its own `properties` definition:
+
+1.  The keys are base URI's to a property type
+1.  The values are defined by either:
+    1.  A reference to a property type
+    1.  An array definition, where
+        1.  the `items` are defined by a reference to a property type
+        1.  `minItems` and/or `maxItems` are optionally defined
+1.  The required fields are defined by a `required` list where the elements are base URIs that are a subset of the keys
+
+A property type Object `X` is therefore compatible with another property type Object `Y` if and only if
+
+1.  Each key in the `properties` of `X` is also in the `properties` of `Y`
```

Comment:
Is this the case for properties which are not required?
It seems to me that if the `properties` for `X` referred to here are not present in the `required` list, they do not need to be present in `Y` for compatibility. 

Or at least we need the inverse relationship as well to make it clear that all optional values need to be present on both sides

Reply:
It is the case. Because if a property in `X` is not required, it is still possible.

Lets say we have a schema `X` with some properties and an optional property `Foo`. Now let's say another schema `Y` has all the properties, but does not have `Foo`.

{
  "foo": ...,
  ...
}

validates against `X` but does not validate against `Y`.

Reply:
Fair enough, just wanted to make sure :)